src/jdk.incubator.code/share/classes/jdk/incubator/code/writer/OpBuilder.java

```diff
@@ -112,54 +120,60 @@ public class OpBuilder {
             J_U_MAP,
             J_U_LIST);
 
-    static final FunctionType BUILDER_F_TYPE = functionType(type(Op.class),
-            type(OpFactory.class),
-            type(TypeElementFactory.class));
+    static final FunctionType BUILDER_F_TYPE = functionType(type(Op.class));
 
 
-    Map<Value, Value> valueMap;
+    final Map<Value, Value> valueMap;
 
-    Map<Block, Value> blockMap;
+    final Map<Block, Value> blockMap;
 
-    Map<ExternalizedTypeElement, Value> exTypeElementMap;
+    final Map<ExternalizedTypeElement, Value> exTypeElementMap;
 
-    Map<TypeElement, Value> typeElementMap;
+    final Map<TypeElement, Value> typeElementMap;
 
-    Block.Builder builder;
+    final Block.Builder builder;
 
-    Value opFactory;
+    final Value dialectFactory;
 
-    Value typeElementFactory;
+    final Value opFactory;
+
+    final Value typeElementFactory;
 
     /**
      * Transform the given code model to one that builds it.
+     * <p>
+     * This method initially applies the function {@code dialectFactoryF} to
+     * the block builder that is used to build resulting code model. The result
+     * is a dialect factory value which is subsequently used to build operations
+     * that construct type elements and operations present in the given code model.
      *
-     * @param op the code model
+     * @param op the code model.
+     * @param dialectFactoryF a function that builds code items to produce a dialect factory value.
      * @return the building code model.
      */
-    public static FuncOp createBuilderFunction(Op op) {
-        return new OpBuilder().build(op);
+    public static FuncOp createBuilderFunction(Op op, Function<Block.Builder, Value> dialectFactoryF) {
```

Comment:
Nice -- the functional parameter effectively allows us to make the compiler-generated op method parameter-less, which then simplifies a lot of other stuff (compiler support and lambda metafactory) -- in reality we could have done this even w/o the dialect factory (e.g. take two lambdas) but good to see this now.